hw/src/ledger.rs

```diff
@@ -365,18 +255,167 @@ fn try_connect_polling(ledger: Arc<Manager>, timeout: Duration) -> bool {
+
+		let eth_path = &ETH_DERIVATION_PATH_BE[..];
+		let etc_path = &ETC_DERIVATION_PATH_BE[..];
+		let derivation_path = match *self.key_path.read() {
+			KeyPath::Ethereum => eth_path,
+			KeyPath::EthereumClassic => etc_path,
+		};
+		const MAX_CHUNK_SIZE: usize = 255;
+		let mut chunk: [u8; MAX_CHUNK_SIZE] = [0; MAX_CHUNK_SIZE];
+		&mut chunk[0..derivation_path.len()].copy_from_slice(derivation_path);
+		let mut dest_offset = derivation_path.len();
+		let mut data_pos = 0;
+		let mut result;
+		loop {
+			let p1 = if data_pos == 0 { 0x00 } else { 0x80 };
+			let dest_left = MAX_CHUNK_SIZE - dest_offset;
+			let chunk_data_size = min(dest_left, transaction.len() - data_pos);
+			&mut chunk[dest_offset..][0..chunk_data_size].copy_from_slice(&transaction[data_pos..][0..chunk_data_size]);
+			result = Self::send_apdu(&handle, commands::SIGN_ETH_TRANSACTION, p1, 0, &chunk[0..(dest_offset + chunk_data_size)])?;
+			dest_offset = 0;
+			data_pos += chunk_data_size;
+			if data_pos == transaction.len() {
+				break;
+			}
+		}
+
+		if result.len() != 65 {
+			return Err(Error::Protocol("Signature packet size mismatch"));
+		}
+		let v = (result[0] + 1) % 2;
+		let r = H256::from_slice(&result[1..33]);
+		let s = H256::from_slice(&result[33..65]);
+		Ok(Signature::from_rsv(&r, &s, v))
+	}
+
+	fn set_key_path(&self, key_path: KeyPath) {
+		*self.key_path.write() = key_path;
+	}
+
+	fn update_devices(&self) -> Result<usize, Self::Error> {
+		let mut usb = self.usb.lock();
+		usb.refresh_devices();
+		let devices = usb.devices();
+		let mut new_devices = Vec::new();
+		let mut num_new_devices = 0;
+		for device in devices {
+			trace!("Checking device: {:?}", device);
+			if device.vendor_id != LEDGER_VID || !LEDGER_PIDS.contains(&device.product_id) {
+				continue;
+			}
+			match self.read_device(&usb, &device) {
+				Ok(info) => {
+					debug!("Found device: {:?}", info);
+					if !self.devices.read().iter().any(|d| d.path == info.path) {
+						num_new_devices += 1;
+					}
+					new_devices.push(info);
+
+				}
+				Err(e) => debug!("Error reading device info: {}", e),
+			};
+		}
+		*self.devices.write() = new_devices;
+		Ok(num_new_devices)
+	}
+
+	fn read_device(&self, usb: &hidapi::HidApi, dev_info: &hidapi::HidDeviceInfo) -> Result<Device, Self::Error> {
+		let handle = self.open_path(|| usb.open_path(&dev_info.path))?;
+		let manufacturer = dev_info.manufacturer_string.clone().unwrap_or("Unknown".to_owned());
+		let name = dev_info.product_string.clone().unwrap_or("Unknown".to_owned());
+		let serial = dev_info.serial_number.clone().unwrap_or("Unknown".to_owned());
+		match self.get_address(&handle) {
+			Ok(Some(addr)) => {
+				Ok(Device {
+					path: dev_info.path.clone(),
+					info: WalletInfo {
+						name: name,
+						manufacturer: manufacturer,
+						serial: serial,
+						address: addr,
+					},
+				})
+			}
+			// This variant is not possible, but the trait forces this return type
+			Ok(None) => Err(Error::Placeholder),
```

Comment:
If it's truly not possible, then I'd suggest calling it something more self-explanatory like `Error::ImpossibleError` (and then we can have a laugh when we see it in production somewhere in a year ;P)